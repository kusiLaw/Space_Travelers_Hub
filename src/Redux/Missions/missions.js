// Action types
const READ = 'spaceTravel/missions/READ';
const BOOK = 'spaceTravel/missions/BOOK';
const LEAVE = 'spaceTravel/missions/LEAVE';

// Action creators
export const read = (missions) => ({
  type: READ,
  payload: missions,
});

export const book = (id) => ({
  type: BOOK,
  payload: id,
});

export const leave = (id) => ({
  type: LEAVE,
  payload: id,
});

// Reducer
const missionsReducer = (state = [], action) => {
  let newState;
  switch (action.type) {
    case READ:
      return action.payload;
    case BOOK:
      newState = state.map((item) => {
        if (item.mission_id !== action.payload) return item;
        return { ...item, reserved: true };
      });
      return newState;
    case LEAVE:
      newState = state.map((item) => {
        if (item.mission_id !== action.payload) return item;
        return { ...item, reserved: false };
      });
      return newState;
    default:
      return state;
  }
};

export const recieveMissions = () => async (dispatch) => {
  await fetch('https://api.spacexdata.com/v3/missions')
    .then((res) => res.json())
    .then((missions) => {
      const missionsList = [];
      Object.keys(missions).map((key) => {
        missionsList.push({
          mission_id: missions[key].mission_id,
          missionName: missions[key].mission_name,
          description: missions[key].description,
          reserved: false,
        });
        return missionsList;
      });
      dispatch(read(missionsList));
    });
};

export default missionsReducer;
