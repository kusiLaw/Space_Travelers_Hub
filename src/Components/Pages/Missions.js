import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Table from 'react-bootstrap/Table';
import { recieveMissions } from '../../Redux/Missions/missions';
import MissionItem from './MissionItem';

function Missions() {
  const missions = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    if (missions.length === 0) dispatch(recieveMissions());
  });

  return (
    <div>
      <hr />
      <Table responsive striped bordered hover>
        <thead>
          <tr>
            <th style={{ width: '10%' }}>Mission</th>
            <th style={{ width: '60%' }}>Description</th>
            <th style={{ width: '15%' }}>Status</th>
            <th style={{ width: '15%' }}>empty</th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission) => (
            <MissionItem
              key={mission.mission_id}
              mission_id={mission.mission_id}
              missionName={mission.missionName}
              description={mission.description}
              reserved={mission.reserved}
            />

          ))}
        </tbody>
      </Table>
    </div>
  );
}

export default Missions;
