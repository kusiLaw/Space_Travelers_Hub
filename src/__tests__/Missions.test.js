/* eslint-disable no-undef */
import { Provider } from 'react-redux';
import Missions from '../Components/Pages/Missions';
import store from '../Redux/configureStore';

const ReactTestRenderer = require('react-test-renderer');

it('Testing missions component:', () => {
  const tree = ReactTestRenderer.create(
    <Provider store={store}>
      <Missions />
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
