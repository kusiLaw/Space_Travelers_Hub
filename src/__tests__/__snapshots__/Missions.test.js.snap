// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing missions component: 1`] = `
<div>
  <hr />
  <div
    className="table-responsive"
  >
    <table
      className="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th
            style={
              Object {
                "width": "10%",
              }
            }
          >
            Mission
          </th>
          <th
            style={
              Object {
                "width": "60%",
              }
            }
          >
            Description
          </th>
          <th
            style={
              Object {
                "width": "15%",
              }
            }
          >
            Status
          </th>
          <th
            style={
              Object {
                "width": "15%",
              }
            }
          >
            empty
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
